modules/spechla/1.0/spechla.smk

```diff
@@ -0,0 +1,310 @@
+    group: 
+        "symlink"
+    run:
+        op.absolute_symlink(input.bam, output.bam)
+        op.absolute_symlink(input.bai, output.bai)
+        op.absolute_symlink(input.bai, output.crai)
+
+
+# Example variant calling rule (multi-threaded; must be run on compute server/cluster)
+rule _spechla_setup:
+    output:
+        txt = touch(CFG["dirs"]["inputs"] + "SpecHLA/clone.done")
+    log:
+        stdout = CFG["logs"]["inputs"] + "git_clone.stdout.log"
+    params:
+        scriptdir = CFG["dirs"]["inputs"] + "SpecHLA"
+    shell:
+        op.as_one_line("""
+        git clone git@github.com:deepomicslab/SpecHLA.git --depth 1  {params.scriptdir} > {log.stdout} 2>&1 && 
+        chmod +x -R {params.scriptdir}/bin/*
+        """)
+
+rule _spechla_index:
+    input:
+        txt = CFG["dirs"]["inputs"] + "SpecHLA/clone.done"
+    output:
+        txt = touch(CFG["dirs"]["inputs"] + "SpecHLA/index.done")
+    log:
+        stdout = CFG["logs"]["inputs"] + "index.stdout.log"
+    params:
+        scriptdir = CFG["dirs"]["inputs"] + "SpecHLA"
+    threads:
+        CFG["threads"]["setup"]
+    resources:
+        **CFG["resources"]["setup"]  
+    shell:
+        op.as_one_line("""
+        unset LD_LIBRARY_PATH && unset LIBRARY_PATH &&
+        bash {params.scriptdir}/index.sh > {log.stdout} 2>&1
+        """)
+
+# Extract HLA reads from bam and cram
+def _which_genome(wildcards):
+    if "38" in str({wildcards.genome_build}):
+        return "hg38"
+    else:
+        return "hg19"
+
+rule _spechla_extract_reads:
+    input:
+        txt = CFG["dirs"]["inputs"] + "SpecHLA/index.done",
+        bam = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bam",
+        bai = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.bai",
+        crai = CFG["dirs"]["inputs"] + "bam/{seq_type}--{genome_build}/{sample_id}.crai"
+    output:
+        fastq_1 = temp(CFG["dirs"]["hla_reads"] + "{seq_type}--{genome_build}/{sample_id}_extract_1.fq.gz"),
+        fastq_2 = temp(CFG["dirs"]["hla_reads"] + "{seq_type}--{genome_build}/{sample_id}_extract_2.fq.gz"),
+        fastq_unpaired = temp(CFG["dirs"]["hla_reads"] + "{seq_type}--{genome_build}/{sample_id}_extract.unpaired.fq.gz"),
+    log:
+        stderr = CFG["logs"]["hla_reads"] + "{seq_type}--{genome_build}/{sample_id}/extract_reads.stderr.log"
+    params:
+        scriptdir = CFG["dirs"]["inputs"] + "SpecHLA",
+        sample_id = "{sample_id}",
+        genome = _which_genome,
+        outdir = CFG["dirs"]["hla_reads"] + "{seq_type}--{genome_build}/"
+    conda: CFG["conda_envs"]["spechla"]
+    group: "extract_and_run"
+    threads:
+        CFG["threads"]["extract_reads"]
+    resources:
+        **CFG["resources"]["extract_reads"]  
+    shell:
+        op.as_one_line("""
+        bash {params.scriptdir}/script/ExtractHLAread.sh -s {params.sample_id} -b {input.bam} -r {params.genome} -o {params.outdir} > {log.stderr} 2>&1
```

Comment:
Does not really matter but if you change here the `{params.sample_id}` to `{wildcards.sample_id}` there will be no need to define a parameter for this :smile:

Reply:
Changed all to wildcards.sample_id